File: erettsegi/4_Tesztverseny/tesztverseny.pas
//Emelt érettségi 2017 május, https://www.informatikatanarok.hu/media/uploads/Informatika_erettsegi/Emelt/e_inf_17maj_fl.pdf
program tesztverseny;
uses crt, sysutils, strutils;

const max = 500;
	  adatok = 'valaszok.txt';
	  
Type Sorok = record
	Azonosito:string[5];
	Valasz:string[14];
	Pontszam:integer;
END;

var fajl:text;
	Jovalasz:string[14];
	Lista:array[1..500] of Sorok;
	ossz:integer = 0;

Procedure Beolvasas();
var sor:string;
	darab:array[1..2] of string;
	i:integer;
BEGIN
	WriteLn('1. feladat: Az adatok beolvasása');
	Assign(fajl, adatok);
	Reset(fajl);
	
	ReadLn(fajl, Jovalasz);
	
	While NOT EOF(fajl) do BEGIN
		ReadLn(fajl, sor);
		for i := 1 to 2 do BEGIN
			darab[i] := ExtractWord(i, sor, [' ']);
		END;
		
		ossz := ossz + 1;
		Lista[ossz].Azonosito := darab[1];
		Lista[ossz].Valasz := darab[2];
	END;
	
	Close(fajl);
END;

Procedure Feladat2();
BEGIN
	WriteLn('2. feladat: A vetélkedőn ', ossz, ' versenyző indult. ');
END;

Procedure Feladat5();
var i:integer;
	sorszam:integer;
	db:integer = 0;
BEGIN
	WriteLn;
	Write('5. feladat: A feladat sorszáma = ');
	ReadLn(sorszam);
	
	for i := 1 to ossz do BEGIN
		if(Lista[i].Valasz[sorszam] = Jovalasz[sorszam]) then BEGIN
			db := db + 1;
		END;
	END;
	
	WriteLn('A feladatra ', db, ' fő, a versenyzők ', ((100/ossz)*db):2:2, '%-a adott helyes választ.');
END;

Procedure Feladat4(index:integer);
var i:integer;
BEGIN
	WriteLn;
	WriteLn('4. feladat: ');
	
	WriteLn('BCCCDBBBBCDAAA (a helyes megoldás)');
	for i := 1 to 14 do BEGIN
		if(Lista[index].Valasz[i] = Jovalasz[i]) then BEGIN
			Write('+');
		END else BEGIN
			Write(' ');
		END;
	END;
	
	Write(' (a versenyző helyes válaszai)');
END;

Procedure Feladat3();
var azon:string[5];
	i:integer;
BEGIN
	Write('3. feladat: A versenyző azonosítója = ');
	ReadLn(azon);
	
	for i := 1 to ossz do BEGIN
		if(Lista[i].Azonosito = azon) then BEGIN
			Write(Lista[i].Valasz, ' (a versenyző válasza)');
			WriteLn;
			Feladat4(i);
		END;
	END;
END;

Function Pontok(fel:integer): integer;
var pont:integer;
BEGIN
	Case fel Of
		1..5: pont := 3;
		6..10: pont :=4;
		11..13: pont := 5;
		14: pont := 6
	END;
	
	Pontok := pont;
END;

Procedure Feladat6();
var i:integer;
	k:integer;
	osszpont:integer = 0;
BEGIN
	WriteLn('6. feladat: A versenyzők pontszámának meghatározása');
	Assign(fajl, 'pontok.txt');
	ReWrite(fajl);
	
	for i := 1 to ossz do BEGIN
		osszpont := 0;
		for k := 1 to 14 do BEGIN
			if(Lista[i].Valasz[k] = Jovalasz[k]) then BEGIN
				osszpont := osszpont + Pontok(k);
			END;
		END;
		
		WriteLn(fajl, Lista[i].Azonosito, ' ', osszpont);
		Lista[i].Pontszam := osszpont;
	END;
	
	Close(fajl);
END;

Procedure Feladat7();
var i:integer;
	k:integer;
	dijak:integer = 1;
	temp:Sorok;
BEGIN
	WriteLn('7. feladat: A verseny legjobbjai:');
	
	for k := 1 to ossz do BEGIN
		for i := 1 to ossz-k do BEGIN
			if(Lista[i].Pontszam < Lista[i+1].Pontszam) then BEGIN
				temp := Lista[i];
				Lista[i] := Lista[i+1];
				Lista[i+1] := temp;
			END;
		END;
	END;
	
	WriteLn(dijak, '. díj (', Lista[1].Pontszam, ' pont): ', Lista[1].Azonosito);
	
	for i := 2 to ossz do BEGIN
		if(Lista[i].Pontszam < Lista[i-1].Pontszam) then BEGIN
			dijak := dijak + 1;
		END;
		
		if(dijak <= 3) then BEGIN
			WriteLn(dijak, '. díj (', Lista[i].Pontszam, ' pont): ', Lista[i].Azonosito);
		END;
	END;
END;

BEGIN
	ClrScr;
	
	Beolvasas();
	WriteLn;
	Feladat2();
	WriteLn;
	Feladat3();
	WriteLn;
	Feladat5();
	WriteLn;
	Feladat6();
	WriteLn;
	Feladat7;
	
	WriteLn;
	WriteLn;
END.

File: erettsegi/4_Tesztverseny/tesztverseny.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace tesztverseny
{
    class tesztverseny
    {
        public class Versenyzo
        {
            public string Azonosito;
            public string Valaszok;
            public int Pontszam;

            public Versenyzo() { }
        }

        static List<Versenyzo> Versenyzok = new List<Versenyzo>();
        static string Valaszok;

        static void Main(string[] args)
        {
            Feladat1();
            Feladat2();
            Feladat3();
            Feladat5();
            Feladat6();
            Feladat7();
        }

        static void Feladat1()
        {
            Console.WriteLine("1. feladat: Az adatok beolvasása\n");
            FileStream fs = File.Open("valaszok.txt", FileMode.Open);
            StreamReader sr = new StreamReader(fs);

            Valaszok = sr.ReadLine();
            while(!sr.EndOfStream)
            {
                string[] darab = sr.ReadLine().Split(' ');
                Versenyzok.Add(new Versenyzo
                {
                    Azonosito = darab[0],
                    Valaszok = darab[1],
                    Pontszam = 0
                });
            }

            sr.Close();
            fs.Close();
        }

        static void Feladat2()
        {
            Console.WriteLine($"2. feladat: A vetélkedőn {Versenyzok.Count} versenyző indult.\n");
        }

        static void Feladat3()
        {
            Console.Write("3. feladat: A versenyző azonosítója = ");
            string azonosito = Console.ReadLine();

            Versenyzo versenyzo = Versenyzok.Find(x => x.Azonosito == azonosito);
            Console.WriteLine($"{versenyzo.Valaszok} (a versenyző válasza)\n");
            Feladat4(versenyzo); //Az is elég ha csak a válaszait adod tovább, nem az egész versenyzőt
            
            /* vagy
            foreach(Versenyzo i in Versenyzok)
            {
                if(i.Azonosito == azonosito)
                {
                    Console.WriteLine($"{i.Valaszok} (a versenyző válasza)");
                    Feladat4(i);
                    break;
                }
            }
            */
        }

        static void Feladat4(Versenyzo versenyzo)
        {
            Console.WriteLine(Valaszok);
            for(int i = 0; i < 14; i++)
            {
                Console.Write(versenyzo.Valaszok[i] == Valaszok[i] ? "+" : " ");
            }

            /* vagy
            for(int i = 0; i < 14; i++)
            {
                if(versenyzo.Valaszok[i] == Valaszok[i])
                {
                    Console.Write("+");
                } else
                {
                    Console.Write(" ");
                }
            }
            */
        }

        static void Feladat5()
        {
            Console.Write("\n\n5. feladat: A feladat sorszáma = ");
            int feladat = Convert.ToInt32(Console.ReadLine());

            int darab = 0;
            foreach(Versenyzo i in Versenyzok)
            {
                if (i.Valaszok[feladat] == Valaszok[feladat])
                    darab++;
            }

            Console.WriteLine($"A feladatra {darab} fő, a versenyzők {(darab / Versenyzok.Count) * 100.0f}%-a adott helyes választ.\n");
        }

        static void Feladat6()
        {
            Console.WriteLine("6. feladat: A versenyzők pontszámának meghatározása\n");

            FileStream fs = File.Open("pontok.txt", FileMode.Append);
            StreamWriter sw = new StreamWriter(fs);

            foreach(Versenyzo versenyzo in Versenyzok)
            {
                for(int i = 0; i < 14; i++)
                {
                    if(versenyzo.Valaszok[i] == Valaszok[i])
                    {
                        if (i <= 5) versenyzo.Pontszam += 3;
                        else if (i <= 10) versenyzo.Pontszam += 4;
                        else if (i <= 13) versenyzo.Pontszam += 5;
                        else versenyzo.Pontszam += 6;
                    }
                }

                sw.WriteLine($"{versenyzo.Azonosito} {versenyzo.Pontszam}");
            }

            sw.Close();
            fs.Close();
        }

        static void Feladat7()
        {
            Console.WriteLine("7. feladat: A verseny legjobbjai:");

            for(int i = 1; i < Versenyzok.Count; i++)
            {
                for(int k = 0; k < Versenyzok.Count - i; k++)
                {
                    if(Versenyzok[k].Pontszam < Versenyzok[k + 1].Pontszam)
                    {
                        Versenyzo temp = Versenyzok[k + 1];
                        Versenyzok[k + 1] = Versenyzok[k];
                        Versenyzok[k] = temp;
                    }
                }
            }

            int dij = 1;
            Console.WriteLine($"1. díj ({Versenyzok[0].Pontszam} pont): {Versenyzok[0].Azonosito}");

            for (int i = 1; i < Versenyzok.Count; i++)
            {
                if (Versenyzok[i].Pontszam > Versenyzok[i + 1].Pontszam)
                {
                    ++dij;
                    if (dij == 4)
                        break;
                }

                Console.WriteLine($"{dij}. díj ({Versenyzok[i].Pontszam} pont): {Versenyzok[i].Azonosito}");
            }
        }
    }
}
